statistics/histogram.go

```diff
@@ -462,30 +451,76 @@ func (hg *Histogram) greaterRowCount(value types.Datum) float64 {
 	return math.Max(0, gtCount)
 }
 
+// locateBucket locates where a value falls in the range of the Histogram.
+// Return value:
+// 	exceed: if the value is larger than the upper bound of the last Bucket of the Histogram
+// 	bucketIdx: assuming exceed if false, which Bucket does this value fall in (note: the range before a Bucket is also
+//		considered belong to this Bucket)
+// 	inBucket: assuming exceed if false, whether this value falls in this Bucket, instead of falls between
+//		this Bucket and the previous Bucket.
+// 	matchLastValue: assuming inBucket is true, if this value is the last value in this Bucket, which has a counter (Bucket.Repeat)
+// Examples:
+// 	val0 |<-[bkt0]->| |<-[bkt1]->val1(last value)| val2 |<--val3--[bkt2]->| |<-[bkt3]->| val4
+// 	locateBucket(val0): false, 0, false, false
+// 	locateBucket(val1): false, 1, true, true
+// 	locateBucket(val2): false, 2, false, false
+// 	locateBucket(val3): false, 2, true, false
+// 	locateBucket(val4): true, 3, false, false
+func (hg *Histogram) locateBucket(value types.Datum) (exceed bool, bucketIdx int, inBucket, matchLastValue bool) {
+	// Empty histogram
+	if hg == nil || hg.Bounds.NumRows() == 0 {
+		return true, 0, false, false
+	}
+	index, match := hg.Bounds.LowerBound(0, &value)
+	// The value is larger than the max value in the histogram (exceed is true)
+	if index >= hg.Bounds.NumRows() {
+		return true, hg.Len() - 1, false, false
+	}
+	bucketIdx = index / 2
+	// The value is before this bucket
+	if index%2 == 0 && !match {
+		return false, bucketIdx, false, false
+	}
+	// The value matches the last value in this bucket
+	if (index%2 == 1 || hg.bucketIsPoint(bucketIdx)) && match {
```

Comment:
It seems we do not need the function `bucketIsPoint`. We can compare the `value` to the last value of the bucket directly.

Reply:
Yes, you are right!
Updated.